ReactWindows/ReactNative/Views/Image/ReactImageManager.cs

```diff
@@ -303,13 +332,91 @@ private void OnImageStatusUpdate(Border view, ImageLoadStatus status, ImageMetad
         /// <param name="view">The image view instance.</param>
         private void SetUriFromMultipleSources(Border view)
         {
-            var sources = default(List<KeyValuePair<string, double>>);
-            if (_imageSources.TryGetValue(view.GetTag(), out sources))
+            var sources = GetImageData(view).Sources;
+            if (sources != default(List<KeyValuePair<string, double>>))
             {
                 var targetImageSize = view.Width * view.Height;
                 var bestResult = sources.LocalMin((s) => Math.Abs(s.Value - targetImageSize));
                 SetUriFromSingleSource(view, bestResult.Key);
             }
         }
+
+        private ReactImageData GetImageData(Border view)
+        {
+            var tag = view.GetTag();
+            var data = default(ReactImageData);
+            if (!_imageData.TryGetValue(tag, out data))
+            {
+                data = new ReactImageData(view);
+                _imageData.Add(tag, data);
+            }
+
+            return data;
+        }
+
+        class ReactImageData : IDisposable
+        {
+            private List<KeyValuePair<string, double>> _sources;
+
+            public List<KeyValuePair<string, double>> Sources
+            {
+                get
+                {
+                    return _sources;
+                }
+            }
+
+            public string SourceUri { get; set; }
+
+            public EffectManager EffectManager { get; private set; }
+
+            public uint? TintColor { get; set; }
+
+            public double? BlurRadius { get; set; }
+
+            public ReactImageData(FrameworkElement element)
+            {
+                EffectManager = new EffectManager(element);
+            }
+
+            public void SetSources(JArray sources)
+            {
+                if (_sources != default(List<KeyValuePair<string, double>>))
+                {
+                    _sources.Clear();
+                }
+                else
+                {
+                    _sources = new List<KeyValuePair<string, double>>(sources.Count);
+                }
+
+                foreach (var source in sources)
+                {
+                    var sourceData = (JObject)source;
+                    _sources.Add(
+                        new KeyValuePair<string, double>(
+                            sourceData.Value<string>("uri"),
+                            sourceData.Value<double>("width") * sourceData.Value<double>("height")));
+                }
+
+                _sources.Sort((p1, p2) => p1.Value.CompareTo(p2.Value));
+            }
+
+            protected virtual void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    EffectManager.Dispose();
+                }
+            }
```

Comment:
I think you only need to implement this pattern if the disposable type is not sealed. You could save 7 lines of code by sealing ReactImageData and calling EffectManager.Dispose() directly in the Dispose() method.

Reply:
For that matter, the class is private inside ReactImageManager, so it doesn't even need to be sealed. You can just delete the Dispose(bool) stuff.

---
In reply to: [136225213](https://github.com/Microsoft/react-native-windows/pull/1331#discussion_r136225213) [](ancestors = 136225213)

Reply:
Agree, totally forgot that I've just made it as internal class. Will be fixed.